### DISCUSSION TITLE
Crazyflie freezes when running app layer apps

### DISCUSSION THREAD

**Discussion by quinnk4** [bitcraze/discussions#613] [Q&A] [answered]
I have written a few C apps using the application layer, and all of them (including the default examples) run into a problem where the program will run fine for a certain amount of time (a few seconds up to a minute) then freeze. The program will start running again if I connect to the drone via CFClient, but if I disconnect again the same thing happens.

I wrote a simple program using the LED ring that flashes on and off, to indicate whether the app is still running or not. What this should do is flash red once every second continuously, but after a while it freezes and gets stuck with all LEDs either on red or off.

`#include <stdbool.h>`
`#include "app.h"`
`#include "FreeRTOS.h"`
`#include "task.h"`
`#include "headerInOtherDir.h"`
`#include "param.h"`
``
`void appMain()`
`{`
`  paramVarId_t ledeffect = paramGetVarId("ring", "effect");`
`  paramVarId_t Red = paramGetVarId("ring", "solidRed");`
`  paramVarId_t Blue = paramGetVarId("ring", "solidBlue");`
`  paramVarId_t Green = paramGetVarId("ring", "solidGreen");`
`  uint8_t solidColourEffect = 7;`
`  uint8_t high_value = 20;`
`  uint8_t low_value = 0;`
``
`  paramSetInt(ledeffect, solidColourEffect);`
`  paramSetInt(Blue, low_value);`
`  paramSetInt(Green, low_value);`
`  while(1) {`
`    vTaskDelay(M2T(1000));`
`    paramSetInt(Red, high_value);`
`    vTaskDelay(M2T(50));`
`    paramSetInt(Red, low_value);`
`  }`
`}`

There's no other indication of what's happening, the heartbeat LED on the Crazyflie keeps blinking as normal and the two blue LEDs remain on. What could be causing this freezing?

**Comment by ataffanel** [comment, marked as answer]
Are you using the latest version of the firmware?

One thing that could be happening and that would explain your problem perfectly is that `paramSetInt` used to block if a client was not connected: this function sends a change notification via radio and if no client is connected the notification packets will fill up the TX queue and then the function will block.

This has been changed recently so you should make sure that you are running on the latest version from git.

**Reply by quinnk4** [reply]
I'm using the version of the firmware included in the VM, how do I check its version and update it if needed?

**Reply by quinnk4** [reply]
Ok, I think I have it updated now - the submodules in the firmware repository tripped me up a little. It looks like updating the firmware to the latest version fixed the issue, thanks!
